Mage.Sets/src/mage/cards/i/IdolOfEndurance.java

```diff
@@ -0,0 +1,194 @@
+import mage.constants.CardType;
+import mage.constants.ComparisonType;
+import mage.constants.Duration;
+import mage.constants.Outcome;
+import mage.constants.WatcherScope;
+import mage.constants.Zone;
+import mage.filter.FilterCard;
+import mage.filter.StaticFilters;
+import mage.filter.common.FilterCreatureCard;
+import mage.filter.predicate.mageobject.ConvertedManaCostPredicate;
+import mage.game.ExileZone;
+import mage.game.Game;
+import mage.game.events.GameEvent;
+import mage.players.Player;
+import mage.util.CardUtil;
+import mage.watchers.Watcher;
+
+import java.util.Objects;
+import java.util.UUID;
+
+/**
+ *
+ * @author htrajan
+ */
+public final class IdolOfEndurance extends CardImpl {
+
+    public IdolOfEndurance(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.ARTIFACT}, "{2}{W}");
+
+        // When Idol of Endurance enters the battlefield, exile all creature cards with converted mana cost 3 or less from your graveyard until Idol of Endurance leaves the battlefield.
+        Ability ability = new EntersBattlefieldTriggeredAbility(new IdolOfEnduranceExileEffect());
+        ability.addEffect(new CreateDelayedTriggeredAbilityEffect(new OnLeaveReturnExiledAbility(Zone.GRAVEYARD)));
+        this.addAbility(ability);
+
+        // {1}{W}, {T}: Until end of turn, you may cast a creature spell from among the cards exiled with Idol of Endurance without paying its mana cost.
+        Ability activatedAbility = new SimpleActivatedAbility(new IdolOfEnduranceResetWatcherEffect(), new ManaCostsImpl<>("{1}{W}"));
+        activatedAbility.addEffect(new IdolOfEnduranceCastFromExileEffect(ability));
+        this.addAbility(activatedAbility, new IdolOfEnduranceWatcher(ability));
+    }
+
+    private IdolOfEndurance(final IdolOfEndurance card) {
+        super(card);
+    }
+
+    @Override
+    public IdolOfEndurance copy() {
+        return new IdolOfEndurance(this);
+    }
+}
+
+class IdolOfEnduranceExileEffect extends OneShotEffect {
+
+    private static final FilterCard filter = new FilterCreatureCard();
+
+    static {
+        filter.add(new ConvertedManaCostPredicate(ComparisonType.FEWER_THAN, 4));
+    }
+
+    public IdolOfEnduranceExileEffect() {
+        super(Outcome.Exile);
+        staticText = "exile all creature cards with converted mana cost 3 or less from your graveyard until {this} leaves the battlefield";
+    }
+
+    @Override
+    public IdolOfEnduranceExileEffect copy() {
+        return new IdolOfEnduranceExileEffect();
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller != null) {
+            controller.getGraveyard().stream()
+                .map(game::getCard)
+                .filter(Objects::nonNull)
+                .filter(card -> filter.match(card, game))
+                .forEach(card -> controller.moveCards(card, Zone.EXILED, source, game));
```

Comment:
the cards should all be moved at once, so create a `Cards` object, add everything to it and then move that whole collection at once.

Reply:
noted for the future, thanks